File: code/HuaWeiNote/app/src/main/java/com/fan/hwnote/app/model/entity/NoteContent.kt
package com.fan.hwnote.app.model.entity

/**
 * 笔记的完整内容：内容块列表 + 手写笔画列表。
 * 持久化为 content_json 字段（详见 PRD §6.2）。
 */
data class NoteContent(
    val blocks: List<Block>,
    val handwriting: List<Stroke>,
) {
    /**
     * 生成 plain_text（搜索字段）：
     * - TextBlock：取 text，非空才计入
     * - ChecklistBlock：每项 text，非空才计入
     * - ImageBlock：忽略
     * - AudioBlock：忽略
     * - handwriting：忽略
     * 多段用 '\n' 拼接。
     */
    fun toPlainText(): String {
        val sb = StringBuilder()
        for (b in blocks) {
            when (b) {
                is Block.TextBlock -> appendIfNotEmpty(sb, b.text)
                is Block.ChecklistBlock -> b.items.forEach { appendIfNotEmpty(sb, it.text) }
                is Block.ImageBlock -> Unit // 图片不进搜索
                is Block.AudioBlock -> Unit // 音频不进搜索
            }
        }
        return sb.toString()
    }

    private fun appendIfNotEmpty(sb: StringBuilder, text: String) {
        if (text.isEmpty()) return
        if (sb.isNotEmpty()) sb.append('\n')
        sb.append(text)
    }

    companion object {
        fun empty(): NoteContent = NoteContent(emptyList(), emptyList())
    }
}

File: code/HuaweiNoteFlutter/lib/models/note_content.dart
import 'dart:convert';
import 'block.dart';
import 'stroke.dart';

class NoteContent {
  final List<Block> blocks;
  final List<Stroke> handwriting;

  const NoteContent({required this.blocks, required this.handwriting});

  factory NoteContent.empty() => const NoteContent(blocks: [], handwriting: []);

  String toPlainText() {
    final parts = <String>[];
    for (final b in blocks) {
      switch (b) {
        case TextBlock():
          if (b.text.isNotEmpty) parts.add(b.text);
        case ChecklistBlock():
          for (final item in b.items) {
            if (item.text.isNotEmpty) parts.add(item.text);
          }
        case ImageBlock():
          break;
        case AudioBlock():
          break;
      }
    }
    return parts.join('\n');
  }

  String toJson() {
    final root = <String, dynamic>{
      'blocks': blocks.map((b) => b.toJson()).toList(),
      'handwriting': {
        'strokes': handwriting.map((s) => s.toJson()).toList(),
      },
    };
    return jsonEncode(root);
  }

  static NoteContent fromJson(String s) {
    try {
      final root = jsonDecode(s) as Map<String, dynamic>;
      final blocksJson = root['blocks'] as List<dynamic>? ?? [];
      final blocks = blocksJson
          .map((b) => Block.fromJson(b as Map<String, dynamic>))
          .whereType<Block>()
          .toList();
      final hw = root['handwriting'] as Map<String, dynamic>? ?? {};
      final strokesJson = hw['strokes'] as List<dynamic>? ?? [];
      final strokes = strokesJson
          .map((s) => Stroke.fromJson(s as Map<String, dynamic>))
          .whereType<Stroke>()
          .toList();
      return NoteContent(blocks: blocks, handwriting: strokes);
    } catch (_) {
      return NoteContent.empty();
    }
  }
}
